### PR TITLE
Add OTX optimize for visual prompting task

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

This PR includes

- [x] enable `otx optimize` for ptq
- [x] remove unused modules
    - [x] transforms
    - [x] segment anything model
- [x] update config

#### PTQ results
* dataset : WGISD

      | torch | ov IR (fp32) | ov IR (ptq only image encoder)
    -- | -- | -- | --
    mDice | 0.9467 | 0.9442 | 0.9432

* Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz

    throughput | f32 | u8
    -- | -- | --
    image   encoder | 1.84 fps | 1.97 fps
    decoder | 107.90 fps | -

* 11th Gen Intel(R) Core(TM) i9-11900 @ 2.50GHz

    throughput  | f32 | u8
    -- | -- | --
    image   encoder | 0.40 fps | 0.40 fps
    decoder | 84.00 fps | -

** PTQ for decoder (prompt encoder + mask decoder) is not supported because 
1. there are unknown issues when compiling compressed model
2. even if completed, and since decoder is very smaller than image encoder, it's not expected to improve performance significantly

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by sungmanc — DISMISSED** [review]
Thanks for the huge effort

File: src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py [hunk 1]
```diff
@@ -227,16 +227,18 @@ def forward(
...
-            return upscaled_masks, scores, stability_scores, areas, masks
+        # if self.config.model.return_extra_metrics:
+        #     stability_scores = self.calculate_stability_score(
+        #         upscaled_masks, self.config.model.mask_threshold, self.config.model.stability_score_offset
+        #     )
+        #     areas = (upscaled_masks > self.config.model.mask_threshold).sum(-1).sum(-1)
+        #     return upscaled_masks, scores, stability_scores, areas, masks
 
-        return upscaled_masks, scores, masks
+        # return upscaled_masks, scores, masks
```
Comment (jaegukhyun) [inline src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:241]: Could you explain about this change?

File: src/otx/algorithms/visual_prompting/tasks/openvino.py [hunk 2]
```diff
@@ -335,4 +408,80 @@ def optimize(
...
+            raise RuntimeError("PTQ optimize failed, model is None")
+
+        if optimization_type is not OptimizationType.POT:
+            raise ValueError("PTQ is the only supported optimization type for OpenVino models")
+
+        dataset = dataset.get_subset(Subset.TRAINING)
+
+        for i, (name, is_encoder) in enumerate(zip(["image_encoder", "decoder"], [True, False]), 1):
+            if name == "decoder":
+                # TODO (sungchul): quantize decoder, too
```
Comment (jaegukhyun) [inline src/otx/algorithms/visual_prompting/tasks/openvino.py:423]: Why can't we quantize decoder?

**Review by jaegukhyun — COMMENTED** [review]
I added some comments. Do you have plan for adding test for PTQ?

File: src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:241]: When converting pth to IR, inner postprocessing is malfunctioning. It is required to look at, but it seems to take time. So I temporarily blocked related parts.

File: src/otx/algorithms/visual_prompting/tasks/openvino.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/openvino.py:423]: When compiling compressed decoder, I have encountered `vector::_M_fill_insert` error.
I think somewhere is needed to fill a value but finding the location seems to take time.
And as you may know, image encoder is very huge than decoder, so inference bottleneck occurred in image encoder. 
(1.84 fps for image encoder and 107.90 fps for decoder on benchmark_app)
So, we decided that it is okay to only compress image encoder.

File: src/otx/algorithms/visual_prompting/tasks/openvino.py [hunk 3]
```diff
@@ -178,6 +194,63 @@ def await_all(self) -> None:
...
+            )
+
+    def __getitem__(self, index: int):
+        """Get item from dataset."""
+        if self.shuffler is not None:
+            index = self.shuffler[index]
+
+        items = self.dataset[index]
+        images, _, prompts = self.inferencer.pre_process(items)
+        processed_image = ResizeLongestSide.apply_image(images["images"][0], self.target_length).transpose(2, 0, 1)
```
Comment (sovrasov) [inline src/otx/algorithms/visual_prompting/tasks/openvino.py]: Why all this preprocess operations are implicitly added to dataset? They could be moved to pre_process, and executed only when a special flag was passed, like `self.inferencer.pre_process(items, extra_processing=True)`

**Review by sovrasov — COMMENTED** [review]
@sungchul2 do you have any accuracy results obtained with PTQ?

File: src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py (hunk 1 above)
Comment (jaegukhyun) [reply src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:241]: That means OV model's output is quite different with PyTorch model?

**Comment by sungchul1** [comment]
> @sungchul2 do you have any accuracy results obtained with PTQ?

I added PTQ performance table in summary.

**Comment by sovrasov** [comment]
> > @sungchul2 do you have any accuracy results obtained with PTQ?
> 
> I added PTQ performance table in summary.

It seems like quantized model inference speed is almost the same as fp32. Did you try to run the inference on a machine with AVX-512 support?

File: src/otx/algorithms/visual_prompting/tasks/openvino.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/openvino.py]: You mean, the below modification is better instead of using the function from `ResizeLongestSide`, right?

```python
# https://github.com/openvinotoolkit/training_extensions/blob/releases/1.4.0/src/otx/algorithms/visual_prompting/adapters/openvino/model_wrappers/openvino_models.py#L47-L51
def preprocess(self, inputs: np.ndarray, extra_processing: bool = True) -> Tuple[Dict[str, np.ndarray], Dict[str, Any]]:
    """Update meta for image encoder."""
    dict_inputs, meta = super().preprocess(inputs)
    if extra_processing:
        dict_inputs["images"] = self.apply_image(dict_inputs["images"][0], self.target_length).transpose(2, 0, 1)[None]
    meta["resize_type"] = self.resize_type
    return dict_inputs, meta

def apply_image(self, ...):
    ...
```

**Comment by sungchul1** [comment]
> > > @sungchul2 do you have any accuracy results obtained with PTQ?
> > 
> > 
> > I added PTQ performance table in summary.
> 
> It seems like quantized model inference speed is almost the same as fp32. Did you try to run the inference on a machine with AVX-512 support?

Updated the table. As far as I know, those machines support AVX-512.

File: src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:241]: Only `upscaled_masks`. Because post-processing `masks` outside of the model looks fine, I think there is little difference of `masks` between torch and IR.
I think inner post-processing is something wrong, but I don't know the reason.

Comment (jaegukhyun) [reply src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:241]: In conclusion, postprocessing is done outside of model due to malfunctioning inner postprocessing funciton. Right?

Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/adapters/pytorch_lightning/models/visual_prompters/segment_anything.py:241]: Right. These temporary blocks are just to avoid inner post-processed outputs.
outer postprocessing is done at https://github.com/openvinotoolkit/training_extensions/blob/a8c665805ea8cf435ca966fbebf9330dbab68794/src/otx/algorithms/visual_prompting/tasks/openvino.py#L135C1-L141

**Review by jaegukhyun — DISMISSED** [review]
LGTM, When the E2E tests will be added?

**Comment by sungchul1** [comment]
> LGTM, When the E2E tests will be added?

I'll do soon :)

File: src/otx/algorithms/visual_prompting/tasks/openvino.py (hunk 3 above)
Comment (sovrasov) [reply src/otx/algorithms/visual_prompting/tasks/openvino.py]: I don't know about ResizeLongestSide, I said this calls sequence can be moved to the inferencer, since it's also a kind of preprocessing.

Comment (sungchul1) [reply src/otx/algorithms/visual_prompting/tasks/openvino.py]: Done.

**Review by sovrasov — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]
